File: ios/Classes/HeicToJpegPlugin.swift
import Flutter
import UIKit


public class HeicToJpegPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "heic_to_jpeg", binaryMessenger: registrar.messenger())
        let instance = HeicToJpegPlugin()
            registrar.addMethodCallDelegate(instance, channel: channel)
        }


        public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
            switch call.method {
                case "convert":
                    guard let bytes = call.arguments as? FlutterStandardTypedData else {
                        result(FlutterError(code: "bad_args", message: "Expected bytes", details: nil))
                        return
                    }

                    guard let uiImage = UIImage(data: bytes.data) else {
                        result(FlutterError(code: "decode_error", message: "Cannot decode HEIC", details: nil))
                        return
                    }

                    guard let jpegData = uiImage.jpegData(compressionQuality: 0.9) else {
                        result(FlutterError(code: "encode_error", message: "Cannot encode JPEG", details: nil))
                        return
                    }

                    result(FlutterStandardTypedData(bytes: jpegData))


                default:
                    result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/io/appzone/heic_to_jpeg/HeicToJpegPlugin.kt
package io.appzone.heic_to_jpeg

import android.graphics.BitmapFactory
import android.graphics.Bitmap
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.ByteArrayOutputStream


class HeicToJpegPlugin: FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel


    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "heic_to_jpeg")
        channel.setMethodCallHandler(this)
    }


    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "convert" -> {
                val bytes = call.arguments as? ByteArray
                if (bytes == null) {
                    result.error("bad_args", "Expected byte array", null)
                    return
                }
                try {
                    val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                    if (bitmap == null) {
                        result.error("decode_error", "Cannot decode HEIC", null)
                        return
                    }


                    val stream = ByteArrayOutputStream()
                    bitmap.compress(Bitmap.CompressFormat.JPEG, 90, stream)
                    result.success(stream.toByteArray())
                } catch (e: Exception) {
                    result.error("conversion_error", e.message, null)
                }
            }
            else -> result.notImplemented()
        }
    }


    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}